[Doc] Add more steps to getting started tutorial (#206)
liqotech/liqo: docs/pages/User/GettingStarted/peer.md

weight: 2
---

> Once Liqo is installed in your cluster, you can start establishing new *peerings* (see the [introduction](../../liqo-brief/#peering-basics) for a high-level overview of this process).

Reviewer: ```suggestion
Once Liqo is installed in your cluster, you can start establishing new *peerings*; an a high-level overview of this process is available in the [Liqo peering basics](../../liqo-brief/#peering-basics) section).
```

Specifically, you can rely on three different methods to peer with other clusters:

1. **LAN Discovery**: automatic discovery of neighboring clusters available in the same LAN. This looks similar to the automatic discovery of Wi-Fi hotspots and it is particularly suitable when your cluster is composed of a single node (e.g., in combination of [K3s](https://k3s.io)).
2. **DNS Discovery**: automatic discovery of the clusters associated with a specific DNS domain (e.g.; *liqo.io*), by scraping the existence of specific DNS entries. This looks similar to the discovery of voice-over-IP SIP servers and it is mostly oriented to big organizations that wish to adopt Liqo in production.
3. **Manual Configuration**: manual addition of specific clusters to the list of known ones. This method is particularly appropriate for testing purposes, as well as to try Liqo outside LAN, without requiring any DNS configuration.

All the three alternatives can be directly configured via the graphical [Dashboard](../dashboard), by means of the Liqo Home view, as well as through standard `kubectl` commands. Now, just proceed to the section most appropriate for your needs.

## LAN Discovery

> As mentioned above, inside a LAN Liqo is able to automatically discover available clusters, as well as to make your cluster discoverable by others.

Reviewer: ```suggestion
Liqo is able to automatically discover any available clusters running on the same LAN, as well as to make your cluster discoverable by others.
```

The currently available and established peerings can be easily monitored through the dashboard.
Just log in, and the home view will present you the list of "Available Peers" (i.e. all discovered foreign clusters).
By default, Liqo tries to establish peerings with on-LAN remote clusters as soon as they are discovered.
Hence, you may already observe a remote cluster presented inside the "Connected Peers" section of the dashboard.

{{%expand "Using kubectl, you can also manually obtain the list of discovered foreign clusters:" %}}

```bash
kubectl get foreignclusters
NAME                                   AGE
ff5aa14a-dd6e-4fd2-80fe-eaecb827cf55   101m
```

To check whether Liqo is configured to automatically attempt to peer with the foreign cluster,
you can check the `join` property of the specific `ForeignCluster` resource:
```bash
kubectl get foreignclusters ${FOREIGN_CLUSTER} --template={{.spec.join}}
true
```
{{% /expand %}}

## DNS Discovery

The DNS discovery procedure requires two orthogonal actions to be enabled.

> 1. Register your own cluster, to make it discoverable by others.

Reviewer: ```suggestion
1. Register your own cluster into your DNS server to make it discoverable by others (the required parameters are presented in the section below).
```

> 1. Register your own cluster, to make it discoverable by others.

Reviewer: Onestamente sono confuso su questi due step e non sono sicuro che siano ambedue appropriati per la DNS Discovery.

2. Connect to a foreign cluster, specifying the remote domain.

### Register the home cluster

> To let others peer with your clusters, it is necessary to register a set of DNS records, specifying the cluster(s) available in your domain, and the different network parameters required to establish the connection.

Reviewer: ```suggestion
To allow other cluster to peer with your cluster(s), it you need to register a set of DNS records that specify the cluster(s) available in your domain, with the different parameters required to establish the connection.
```

> First, you have to retrieve the Cluster ID (i.e. a unique identifier generated by Liqo for identification purposes during the peering process) of your cluster:

Reviewer: Can this be done through the dashboard?

* The `A` record assigns an IP address to the DNS name of the Kubernetes API server, in this case `1.2.3.4`.
  Although, for the sake of simplicity, this example configures an `A` record, an indirect `CNAME` record would also work.
* the `TXT` record is opaque to the DNS system and it is used to store the `Cluster ID` and the `Liqo Namespace` (i.e. where the Liqo components are installed) parameters associated with the cluster.

{{% /expand %}}

### Connect to a remote cluster

In order to leverage the DNS discovery to peer to a remote cluster, it is necessary to specify the remote domain.
This operation can be easily performed through the graphical dashboard: click on the "+" icon located near Available Peers and then select "Add domain".
Here, you need to configure the following parameters:
1. **Domain**: the domain where the cluster you want to peer with is located.
2. **Name**: a mnemonic name to identify the domain.
3. **Join**: to specify whether to automatically trigger the peering procedure.

> {{%expand "Using kubectl, it is also possible to perform the same configuration." %}}

Reviewer: This looks very good to me: add also kubectl commands in this guide, but hiding the text so that it looks available only to most experienced people.

```
cat << "EOF" | kubectl apply -f
apiVersion: discovery.liqo.io/v1alpha1
kind: SearchDomain
metadata:
  name: remote.com
spec:
  domain: remote.com
  autojoin: true
EOF
```

{{% /expand %}}

## Manual Configuration

If the cluster you want to peer with is not present in your LAN, and you do not want to configure the DNS discovery,
it is possible to set-up a manual peering via the graphical dashboard.

First, you have to collect some information about the remote cluster. In particular:

1. The **API Server address**: the address of the remote cluster. It should be accessible without NAT from the Home cluster.

>     To this end, you need to export the right `KUBECONFIG` variable pointing to the remote cluster and then extract the address of

Reviewer: Shouldn't be possible to see this data from the dashboard?

the Kubernetes API server:
    ```bash
    export KUBECONFIG=$REMOTE_KUBECONFIG_PATH
    kubectl config view -o jsonpath='{.clusters[*].cluster.server}'
    ```
    > *Note:* if you have multiple clusters listed in your kubeconfig, the last command will return the API server address for each of them.

> 2. The **Cluster ID**: the UUID which identifies the remote cluster. It can be obtained through:

Reviewer: Also this from dashboard?

    ```
    export KUBECONFIG=$REMOTE_KUBECONFIG_PATH
    kubectl get configmap -n liqo cluster-id -o jsonpath='{.data.cluster-id}'
    ```

Now, through the dashboard of the home cluster, you can configure the new peering. Click on the "+" icon located near Available Peers and then select "Add Remote Peer". Here, you need to configure the parameters obtained during the previous step. Additionally, you also have to set the following values:

 1. **Name**: Name of the cluster (i.e. a mnemonic name, which can also be identical to the clusterID)
 2. **Discovery Type**: Manual

>  3. **AllowedUntrustedCA**: True 

Reviewer: ```suggestion
 3. **AllowedUntrustedCA**: True (i.e., peer to remote cluster even if its connection is not totally secure, as the cluster name is not signed by a trusted certification authority)
```
